lib/Sema/SemaBounds.cpp

```diff
@@ -578,43 +662,29 @@ namespace {
         case CastKind::CK_ArrayToPointerDecay:
           return LValueBounds(E);
         default:
-          return CreateBoundsNone();
+          return CreateBoundsUnknown();
       }
     }
 
     // Compute the bounds of an expression that produces an rvalue.
     BoundsExpr *RValueBounds(Expr *E) {
-      assert(E->isRValue());
+      if (!E->isRValue()) return CreateBoundsInferenceError();
+
       E = E->IgnoreParens();
-      switch (E->getStmtClass()) {
-        case Expr::IntegerLiteralClass: {
-         IntegerLiteral *Lit = dyn_cast<IntegerLiteral>(E);
-         if (!Lit) {
-           llvm_unreachable("unexpected cast failure");
-           return CreateBoundsNone();
-         }
-         if (Lit->getValue() == 0)
-           return CreateBoundsAny();
-
-         return CreateBoundsNone();
-        }
-        case Expr::DeclRefExprClass: {
-          DeclRefExpr *DR = dyn_cast<DeclRefExpr>(E);
-          if (!DR) {
-            llvm_unreachable("unexpected cast failure");
-            return CreateBoundsNone();
-          }
-          EnumConstantDecl *ECD = dyn_cast<EnumConstantDecl>(DR->getDecl());
-          if (ECD->getInitVal() == 0)
-            return CreateBoundsAny();
 
-          return CreateBoundsNone();
```

Comment:
The two cases that return `bounds(none)` when not a null pointer expression are now caught by the `default` at the end of the block.